### ISSUE TITLE
Can't iterate over all possible values of a table

### ISSUE THREAD

**Issue by duvenaud** [google-research/dex-lang#146]
I'm writing a test to see if probabilities sum to one over the set of all possible inputs.  Normally, one has to write tedious code to iterate over the entire set, but Dex should shine at this!  However, writing e.g.
```seqs = for seq:((Fin 3)=>(Fin 4)). seq```
crashes the interpreter:
```
Unexpected type ((Fin 3)=>(Fin 4))
CallStack (from HasCallStack):
  error, called at src/lib/Embed.hs:600:8 in dex-0.1.0.0-WMAsfoEK61B3EtpA608ix:Embed
```

Amusingly, this somehow corrupted the interpreter which then printed
```
UnexpectedU nteyxppee c(t(eFdi nt y3p)e= >(((FFiinn  43)))=
>C(aFlilnS t4a)c)k
 C(aflrloSmt aHcaks C(aflrloSmt aHcaks)C:a
l l Setrarcokr),: 
c a lelrerdo ra,t  csarlcl/eldi ba/tE msbrecd/.lhisb:/6E0m0b:e8d .ihns :d6e0x0-:08. 1i.n0 .d0e-xW-M0A.s1f.o0E.K06-1WBM3AEstfpoAE6K0681iBx3:EEtmpbAe6d0
8ix:Embed
```

**Comment by dougalm** [comment]
Yes, we don't allow table types as index sets right now. There's no fundamental reason not to, except that the space gets very big very quickly!

The bad error message is for the same reason as [yesterday's issue](https://github.com/google-research/dex-lang/issues/143#issuecomment-654766871). Since rewrite #101 we're not checking index set type classes.

My guess for the corruption is that we're writing past the end of an array somewhere, which is overwriting the in-memory program. Is it reproducible?

**Comment by duvenaud** [comment]
Okay, well it's not blocking me.  But if it works someday, it'll be a cute example of the benefits of being able to compose all of Dex's abstractions arbitrarily.

As for reproducing the memory corruption:  It happened consistently yesterday but I can't trigger it again today.

**Comment by duvenaud** [comment]
Closed by #876.
